### PR TITLE
Add nesting support for progress bars

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
- Add `ProgressBarGroup` to store nested progress bars
- Add `contextvar` to `progress` to track reference to current `ProgressBarGroup`
- Added tests for nesting and an example

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Must merge #2580 first.

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @DragaDoncila, could you add a couple of screenshots? Thanks :)

**Comment by DragaDoncila** [comment]
@goanpeca this makes no changes to styling at all, so I don't know if a screenshot would be meaningful - it's just ensuring that nested bars appear next to each other in the dock. I'll update #2656 with a screenshot of what nested bars look like.

**Comment by goanpeca** [comment]
>it's just ensuring that nested bars appear next to each other in the dock.

That sounds meaningful :) 

>I'll update #2656 with a screenshot of what nested bars look like.

Or here 🤷🏼 ?

**Comment by DragaDoncila** [comment]
@goanpeca This is what it will look like once we merge this PR and the styling PR.

 
nested

**Review by goanpeca — APPROVED** [review]
Thanks for working on this @DragaDoncila 😄 

🚀

**Review by sofroniewn — APPROVED** [review]
LGTM too! This has a couple conflicts now, not sure if it will also conflict with #2655, we can merge that one first and then this one.

I'll wait for @tlambert03 to review too though before merge.

Very excited for this to be making such good progress 😉

File: examples/progress_bar_minimal.py
Comment (tlambert03) [inline examples/progress_bar_minimal.py:42]: btw, we've still got some work to do on cleanup here... If you run any of these examples and quit napari while one of these is running, you'll get:

```pytb
napari/napari/_qt/qprogress.py in __exit__(self=<napari._qt.qprogress.progress object>, exc_type=<class 'RuntimeError'>, exc_value=RuntimeError('wrapped C/C++ object of type QLabel has been deleted'), traceback=<traceback object>)
    113         if self.has_viewer:
    114             CURRENT_GROUP.reset(self._group_token)
--> 115             self._pbar.parentWidget().close()
        self._pbar.parentWidget.close = undefined
    116         return super().__exit__(exc_type, exc_value, traceback)
    117

RuntimeError: wrapped C/C++ object of type ProgressBar has been deleted
```

File: napari/_qt/widgets/qt_progress_bar.py
Comment (tlambert03) [inline napari/_qt/widgets/qt_progress_bar.py]: I'd make `current_group_ref` an optional kwarg here.  It's nice to be able to use `get_pbar` without needing to come up with a group ref if you don't care about nesting.

File: napari/_qt/widgets/qt_progress_bar.py
Comment (tlambert03) [inline napari/_qt/widgets/qt_progress_bar.py]: I assume nesting pbars is not going to be nearly as common as just getting a single pbar... so creating a `ProgressBarGroup` for every pbar (just in case it eventually needs to add more) feels a little funny to me.  The alternative would be to make the nested QWidget only at the moment a second one comes along and wants to be nested with it (and reparent the first widget at that time).

File: napari/_qt/widgets/qt_progress_bar.py
Comment (tlambert03) [inline napari/_qt/widgets/qt_progress_bar.py:53]: minor aesthetic changes:  anytime you nest a widget with a layout inside another widget with a layout, the padding starts to really add up.  So you might want to use something like `pbr_group_layout.setContentsMargins(0, 0, 0, 0)` here

also... it'd be really nice to have a border around grouped pbars.  otherwise the fact that they belong together gets lost.  `QFrame` is handy for this purpose (however, I just tried it quickly and the borders aren't showing up... not immediately sure why not here)

File: examples/progress_bar_minimal.py
Comment (tlambert03) [reply examples/progress_bar_minimal.py:42]: actually, If I click the "iterable with context" button twice, I also get this error when the second one finishes.  I think there might be too many things trying to close it (with the `Qt.WA_DeleteOnClose` flag set as well)

Comment (DragaDoncila) [reply examples/progress_bar_minimal.py:42]: Hmmm so our context var approach isn't really detecting nesting specifically, and that's what's leading to these errors I think. 

It's detecting if there's already a context present. So, looking at the minimal examples again, we can click "Iterable" all day long, and that will be fine.

But, if we click "Iterable with Context" twice, our `CURRENT_GROUP` will be set already, and we'll see this as a nested context, even though it's actually just two separate contexts.
This is also true with "Indeterminate" because we're using it in a context.

This is particularly noticeable now that we don't always have a `ProgressBarGroup`. With "Indeterminate", clicking it a second time results in a "ProgressBarGroup" being created. When one of these bars hits 42, it closes the whole group, which also closes the progress bar that was still going.

It feels like with this approach we'd need some sort of GUID (or at least UID) for the context var, so we can detect not only if we've entered the context, but also if we're creating a new progress bar inside the same context. Does this sound sane at all?

Comment (DragaDoncila) [reply examples/progress_bar_minimal.py:42]: > It feels like with this approach we'd need some sort of GUID (or at least UID) for the context var, so we can detect not only if we've entered the context, but also if we're creating a new progress bar inside the same context

Of course, the issue with this approach is we only enter the context after the `ProgressBar`'s been built, so we'd need to move the nesting logic into the `__enter__` method which is potentially quite messy.

Comment (DragaDoncila) [reply examples/progress_bar_minimal.py:42]: Upon further consideration, I don't know if we can achieve what we're trying to achieve here.

The critical issue is that our code is not proceeding procedurally, but is rather event driven. This means that at any point (even while a context is currently being executed), I can click a button and trigger another progress bar to be created.

The creation of this new progress bar could trigger the creation of a new context, and we have no way of knowing whether this new context is an instance of nesting, or just a new context being created somewhere randomly in some code (potentially not even in the same file as our original context).

This means we can't rely on class variables or context variables to tell us whether we're nested, the way we usually might. 

For example, one option I considered was incrementing a count context variable when we first enter a context, and then in future `__enter__`s, if this is >0 then we're nesting, because another context must've been created before us. We can then decrement count in `__exit__`. But this won't work because if I click two buttons in `progress_minimal.py`, and both use contexts, e.g. clicking "Iterable With Context" and "Indeterminate", then checking the count of the context variable will show me 1 - the context of whichever button I clicked first.

I also thought of using `tqdm._instances` - the `WeakSet` used by `tqdm` to track current active objects, but this would fail for the same reason - we can't assume having two active instances of `progress` is due to nesting.

So I'm a little bit stuck here. @tlambert03 any ideas? I feel like the user has to tell us if they're going to be nesting. I can't see another approach...

**Comment by DragaDoncila** [comment]
@tlambert03 I've made good headway I think with getting explicit nesting going. I'm not seeing any `RuntimeErrors` locally, however I did notice if I close the viewer while a progress bar is active it will give that error - same with if we have a test failure. I'm not sure whether that's something we can avoid? Feedback on the approach would be appreciated. 

I'm also not sure why but I couldn't get the `QFrame` separators to show without giving them that object name `QtCustomTitleBar` - I think they must be inheriting a qss property from somewhere. Should I just provide a new name for nested progress bars specifically in the qss?

**Review by sofroniewn — APPROVED** [review]
>  I did notice if I close the viewer while a progress bar is active it will give that error - same with if we have a test failure. I'm not sure whether that's something we can avoid? Feedback on the approach would be appreciated.

I'm not sure how to deal with this, maybe there is a way to shutdown gracefully by connecting into the close signal. Do @tlambert03 or @goanpeca have ideas?

I can also get the same on master right now, so I think it's ok to merge this PR without having that fixed, but would be good to address at some point.

I'm going to move this to the awaiting merge column, but maybe @tlambert03 can give it another pass before merge!! Thanks all for the hard work here!!

**Comment by tlambert03** [comment]
ok... played with this a bit.

The big issue is that the lifetime of the `progress` object, which we've chosen to be the public facing object, can easily "outlive" the lifetime of the viewer.  It is not itself a widget, so unlike most other widgets (which will just go away when the parent closes), the `progress` object will just keep happily iterating even when the viewer is gone.  (for instance, put a print statement in `progress.display`, run the indeterminate example, and quit the viewer in the middle, it will just keep displaying until the thing is done).

The lifetime of the `progress` object needs to be more tightly linked to the lifetime of the viewer.  There's a few options here, but one might be to pass the `progress` instance as an argument to `get_pbar` and then in `get_bar` add something like:
```python
    if progress_iterable is not None:
        viewer_instance.destroyed.connect(progress_iterable.close)
```

The other thing that needs to be done is `self.has_viewer = False` should be set in the `close()` method.  That attribute is only set  in the original __init__ method, but it definitely can change if the viewer is deleted while the iterable is progressing.

If any of that feels a little strange (which I think it kind of does), it's because of the very implicit connection between the `viewer` and the `progress` objects.   Basically, `progress` is just a `tqdm` iterable, which we have not directly connected to the viewer.  So if the viewer closes in the middle of a `with progress()` context, that progress _still_ has to finish itself, and will need to be able to do so without the viewer.

If we want the viewer close even to actually stop the iterable... then we'd need to dig deeper into the `__iter__` method of the `tqdm` class.

**Comment by goanpeca** [comment]
>If any of that feels a little strange (which I think it kind of does), it's because of the very implicit connection between the viewer and the progress objects. Basically, progress is just a tqdm iterable, which we have not directly connected to the viewer. So if the viewer closes in the middle of a with progress() context, that progress still has to finish itself, and will need to be able to do so without the viewer.

Any disadvantage of connecting them?

It does seem like they should be connected. A "headless" mode would still work the same, right?

**Comment by sofroniewn** [comment]
These are great points about shutdown of the progress bar - agreed I think we should try and do something to shutdown a little more gracefully. As nesting actually doesn't make this worse I'm going to merge this now, but maybe @DragaDoncila you can capture this discussion in a new issue which we'll have to work on. Thanks all!!

**Pull request merged**
